Gitlab support

samiyaakhtar opened the pull request:
sample dashboard: http://40.91.86.196:5000/

Closes #141

andrebriggs commented:
@samiyaakhtar the demo looks good. A few questions. 

- When clicking on the sync tag nothing happens, should it link to the git tag in the GitLab repo? 
- Have we tried a pipeline branch other than master? 
- If you were to put an image in your GitLab profile (gravatar) would it show up?
- Currently your GitLab instance is public, what happens if it's private?

samiyaakhtar commented:
@andrebriggs
> When clicking on the sync tag nothing happens, should it link to the git tag in the GitLab repo?

I think this one is a bug. It's also happening on github actions configs but not Azdo. I'll look further

> Have we tried a pipeline branch other than master?
 
I haven't, but I will, perhaps add it to this example itself and test it out. 

> If you were to put an image in your GitLab profile (gravatar) would it show up?

I just uploaded my image to my gitlab account and it is taking a while to start showing up, i'm using the official gitlab avatar API `https://gitlab.com/api/v4/avatar?email=samiyaakhtar7@gmail.com` to fetch the avatar for the user profile. Currently it's still returning my old avatar image, even though I already added an image. Once they update it on their end, it should start showing up.

> Currently your GitLab instance is public, what happens if it's private?

Should still work for a private instance since the configuration requires specifying an access token, but I need to test out this scenario a little further.

evanlouie commented on backend/src/config.ts:
```diff
@@ -13,6 +13,11 @@ export interface IConfig {
   sourceRepo: string;
   hldRepo: string;
 
+  // gitlab
```
Just a styling thing: Since all these are set via `getConfig`, might want to use optional properties instead (i.e. `hldRepo?: string`) then you can don't need all the `||` shortcircuting in `getConfig`

evanlouie commented on backend/src/lib/author.ts:
```diff
@@ -31,27 +32,12 @@ export const get = async (
-        config.sourceRepoAccessToken || config.pipelineAccessToken
-      );
-    }
+    return await fetchAuthor(
```
dont need the `await`

evanlouie commented on backend/src/lib/clustersync.ts:
```diff
@@ -64,6 +74,35 @@ export const get = (): Promise<IClusterSync | undefined> => {
+        });
+    });
   }
   return new Promise((resolve, reject) => {
```
You can just do `return Promise.reject("No tags were found")`

evanlouie commented on packages/spektate/src/IDeployment.ts:
```diff
@@ -433,12 +434,19 @@ export const status = (deployment: IDeployment): string => {
       deployment.hldToManifestBuild.result === "canceled")
   ) {
     return "canceled";
+  } else if ((deployment.srcToDockerBuild &&
```
probably cleaner to do `else if ((deployment. srcToDockerBuild?.result  ?? deployment.dockerToHldReleaseStage?result ?? deployment.hldToManifestBuild?.result) === "success")`

evanlouie commented on packages/spektate/src/pipeline/GitlabPipeline.test.ts:
```diff
@@ -0,0 +1,62 @@
+    expect(Object.keys(pipeline.releases).includes("208859532"));
+    expect(Object.keys(pipeline.releases).includes("208955061"));
+  });
+});
```
newline plz

evanlouie commented on packages/spektate/src/pipeline/GitlabPipeline.ts:
```diff
@@ -0,0 +1,99 @@
+  };
+}
+
+export default GitlabPipeline;
```
newline plz

evanlouie commented on packages/spektate/src/pipeline/mocks/gitlab-pipeline-1.json:
```diff
@@ -0,0 +1,36 @@
+    "illustration": null,
+    "favicon": "https://gitlab.com/assets/ci_favicons/favicon_status_success-8451333011eee8ce9f2ab25dc487fe24a8758c694827a582f17f42b0a90446a2.png"
+  }
+}
```
newline plz

evanlouie commented on packages/spektate/src/pipeline/mocks/gitlab-pipeline-2.json:
```diff
@@ -0,0 +1,36 @@
+    "illustration": null,
+    "favicon": "https://gitlab.com/assets/ci_favicons/favicon_status_success-8451333011eee8ce9f2ab25dc487fe24a8758c694827a582f17f42b0a90446a2.png"
+  }
+}
```
newline plz

evanlouie commented on packages/spektate/src/repository/IGitlabRepo.test.ts:
```diff
@@ -0,0 +1,90 @@
+    expect(url).toBe("https://gitlab.com/samiyaakhtar/hello-world-full-stack/-/tags")
+    jest.spyOn(HttpHelper, "httpGet").mockClear();
+  });
+});
```
newline plz

evanlouie commented on packages/spektate/src/repository/IGitlabRepo.ts:
```diff
@@ -0,0 +1,127 @@
+    return author;
+  }
+  return undefined;
+}
```
newline plz

evanlouie commented on packages/spektate/src/repository/mocks/gitlab-author-response.json:
```diff
@@ -0,0 +1,33 @@
+    "updated_at": "2020-10-28T20:59:09.000Z",
+    "web_url": "https://gitlab.com/samiyaakhtar/hello-world-full-stack/-/pipelines/208927877"
+  }
+}
```
newline plz

evanlouie commented on packages/spektate/src/repository/mocks/gitlab-pr-response.json:
```diff
@@ -0,0 +1,87 @@
+  "user": {
+    "can_merge": true
+  }
+}
```
newline plz

evanlouie commented on packages/spektate/src/repository/mocks/gitlab-releasesurl-response.json:
```diff
@@ -0,0 +1,118 @@
+    },
+    "group_access": null
+  }
+}
```
newline plz

evanlouie commented on packages/spektate/src/repository/mocks/gitlab-tags-response.json:
```diff
@@ -0,0 +1,26 @@
+    "release": null,
+    "protected": false
+  }
+]
```
newline plz

evanlouie reviewed (approved):
lgtm 👍 
Probably want to add an `.editorconfig` to make your editor ensure newlines though

samiyaakhtar commented:
@evanlouie thanks for reviewing! Just added an editorconfig and added newlines to all my files, i think my VS code was removing them before. Also made other changes you suggested

andrebriggs reviewed (approved):
Approved though as said earlier I'm concerned about the dev experience of setting this up. Looking forward to the docs
